ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+   * 2. Deleting all associated index segment and bloom files.
+   * 3. Deleting the clean shutdown file associated with the index of the swap spaces.
+   * 4. Resetting state.
+   * @param recovering {@code true} if this function was called in the context of recovery. {@code false} otherwise.
+   * @throws IOException if there were I/O errors during cleanup.
+   */
+  private void cleanup(boolean recovering) throws IOException {
+    cleanupLogAndIndexSegments(recovering);
+    File cleanShutdownFile = new File(dataDir, TARGET_INDEX_CLEAN_SHUTDOWN_FILE_NAME);
+    if (cleanShutdownFile.exists() && !cleanShutdownFile.delete()) {
+      logger.warn("Could not delete the clean shutdown file {}", cleanShutdownFile);
+    }
+    resetStructures();
+  }
+
+  // copy() helpers
+
+  /**
+   * Sets up the state required for copying data by populating all the data structures, creating a log and index that
+   * wraps the swap spaces and pausing hard delete in the application log.
+   * @throws InterruptedException if hard delete could not be paused.
+   * @throws IOException if logs and indexes could not be set up.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void setupState() throws InterruptedException, IOException, StoreException {
+    CompactionDetails details = compactionLog.getCompactionDetails();
+    long highestGeneration = 0;
+    for (String segmentName : details.getLogSegmentsUnderCompaction()) {
+      highestGeneration = Math.max(highestGeneration, LogSegmentNameHelper.getGeneration(segmentName));
+    }
+    List<Pair<String, String>> targetSegmentNamesAndFilenames = new ArrayList<>();
+    List<LogSegment> existingTargetLogSegments = new ArrayList<>();
+    for (String segmentName : details.getLogSegmentsUnderCompaction()) {
+      long pos = LogSegmentNameHelper.getPosition(segmentName);
+      String targetSegmentName = LogSegmentNameHelper.getName(pos, highestGeneration + 1);
+      String targetSegmentFileName =
+          LogSegmentNameHelper.nameToFilename(targetSegmentName) + TEMP_LOG_SEGMENT_NAME_SUFFIX;
+      File targetSegmentFile = new File(dataDir, targetSegmentFileName);
+      if (targetSegmentFile.exists()) {
+        existingTargetLogSegments.add(new LogSegment(targetSegmentName, targetSegmentFile, metrics));
+      } else {
+        targetSegmentNamesAndFilenames.add(new Pair<>(targetSegmentName, targetSegmentFileName));
+      }
+    }
+    // TODO: available swap space count should be obtained from DiskManager. For now, assumed to be 1.
+    long targetLogTotalCapacity = srcLog.getSegmentCapacity();
+    tgtLog = new Log(dataDir.getAbsolutePath(), targetLogTotalCapacity, srcLog.getSegmentCapacity(), metrics, true,
+        existingTargetLogSegments, targetSegmentNamesAndFilenames.iterator());
+    Journal journal = new Journal(dataDir.getAbsolutePath(), 2 * config.storeIndexMaxNumberOfInmemElements,
```

Comment:
For your first comment, configs changing after restart is handled correctly with no problems. This is the only config value used in this class and is safe for use after restarts.

Emitting metrics is a good point. I have introduced a new `StoreMetrics` class with the store id suffixed by "_temp" for all metrics associated with the target log and index.